File: src/go/upack/list.go
package upack

import (
	"fmt"
	"os"
)

type List struct {
	UserRegistry bool
}

func (*List) Name() string        { return "list" }
func (*List) Description() string { return "Lists packages installed in the local registry." }

func (l *List) Help() string  { return defaultCommandHelp(l) }
func (l *List) Usage() string { return defaultCommandUsage(l) }

func (*List) PositionalArguments() []PositionalArgument {
	return nil
}
func (*List) ExtraArguments() []ExtraArgument {
	return []ExtraArgument{
		{
			Name:        "userregistry",
			Description: "List packages in the user registry instead of the machine registry.",
			Flag:        true,
			TrySetValue: trySetBoolValue("userregistry", func(cmd Command) *bool {
				return &cmd.(*List).UserRegistry
			}),
		},
	}
}

func (l *List) Run() int {
	r := Machine
	if l.UserRegistry {
		r = User
	}

	packages, err := r.ListInstalledPackages()
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		return 1
	}

	for _, pkg := range packages {
		fmt.Println(pkg.groupAndName() + " " + pkg.Version.String())
		if pkg.FeedURL != nil && *pkg.FeedURL != "" {
			fmt.Println("From", *pkg.FeedURL)
		}
		if (pkg.Path != nil && *pkg.Path != "") || pkg.InstallationDate != nil {
			path, date := "<unknown path>", "<unknown date>"
			if pkg.Path != nil && *pkg.Path != "" {
				path = *pkg.Path
			}
			if pkg.InstallationDate != nil {
				date = pkg.InstallationDate.Date.String()
			}
			fmt.Println("Installed to", path, "on", date)
		}
		if (pkg.InstalledBy != nil && *pkg.InstalledBy != "") || (pkg.InstalledUsing != nil && *pkg.InstalledUsing != "") {
			user, application := "<unknown user>", "<unknown application>"
			if pkg.InstalledBy != nil && *pkg.InstalledBy != "" {
				user = *pkg.InstalledBy
			}
			if pkg.InstalledUsing != nil && *pkg.InstalledUsing != "" {
				application = *pkg.InstalledUsing
			}
			fmt.Println("Installed by", user, "using", application)
		}
		if pkg.InstallationReason != nil && *pkg.InstallationReason != "" {
			fmt.Println("Comment:", *pkg.InstallationReason)
		}
		fmt.Println()
	}

	fmt.Println(len(packages), "packages")

	return 0
}

File: src/upack/List.cs
using System.ComponentModel;
using Inedo.UPack.Packaging;

namespace Inedo.UPack.CLI
{
    [DisplayName("list")]
    [Description("Lists packages installed in the local registry.")]
    public sealed class List : Command
    {
        [DisplayName("userregistry")]
        [Description("List packages in the user registry instead of the machine registry.")]
        [ExtraArgument]
        [DefaultValue(false)]
        public bool UserRegistry { get; set; } = false;

        public override async Task<int> RunAsync(CancellationToken cancellationToken)
        {
            IReadOnlyList<RegisteredPackage> packages;
            using (var registry = PackageRegistry.GetRegistry(this.UserRegistry))
            {
                await registry.LockAsync(cancellationToken);
                try
                {
                    packages = await registry.GetInstalledPackagesAsync();
                }
                finally
                {
                    await registry.UnlockAsync();
                }
            }

            foreach (var pkg in packages)
            {
                if (!string.IsNullOrEmpty(pkg.Group))
                {
                    Console.WriteLine($"{pkg.Group}:{pkg.Name} {pkg.Version}");
                }
                else
                {
                    Console.WriteLine($"{pkg.Name} {pkg.Version}");
                }
                if (!string.IsNullOrEmpty(pkg.FeedUrl))
                {
                    Console.WriteLine($"From {pkg.FeedUrl}");
                }
                if (!string.IsNullOrEmpty(pkg.InstallPath) || pkg.InstallationDate != null)
                {
                    Console.WriteLine($"Installed to {(string.IsNullOrEmpty(pkg.InstallPath) ? "<unknown path>" : pkg.InstallPath)} on {(string.IsNullOrEmpty(pkg.InstallationDate) ? "<unknown date>" : pkg.InstallationDate)}");
                }
                if (!string.IsNullOrEmpty(pkg.InstalledBy) || !string.IsNullOrEmpty(pkg.InstalledUsing))
                {
                    Console.WriteLine($"Installed by {(string.IsNullOrEmpty(pkg.InstalledBy) ? "<unknown user>" : pkg.InstalledBy)} using {(string.IsNullOrEmpty(pkg.InstalledUsing) ? "<unknown application>" : pkg.InstalledUsing)}");
                }
                if (!string.IsNullOrEmpty(pkg.InstallationReason))
                {
                    Console.WriteLine($"Comment: {pkg.InstallationReason}");
                }
                Console.WriteLine();
            }

            Console.WriteLine($"{packages.Count} packages");

            return 0;
        }
    }
}
